File: templates/new_node/test/test_service.cpp
// ═══ 纯 C++ 测试：不需要 ROS2，直接测试 Service 层 ═══
#include <gtest/gtest.h>

#include "__PACKAGE_NAME__/model/types.hpp"
#include "__PACKAGE_NAME__/service/example_service.hpp"

using namespace __PACKAGE_NAME__;

class ExampleServiceTest : public ::testing::Test {
protected:
    void SetUp() override {
        model::Config config;
        config.threshold = 0.5;
        service_ = std::make_unique<service::ExampleService>(config);
    }
    std::unique_ptr<service::ExampleService> service_;
};

TEST_F(ExampleServiceTest, NormalInput) {
    model::InputData input{.value = 2.0, .timestamp = 0.0};
    auto output = service_->process(input);
    EXPECT_NEAR(output.result, 1.0, 1e-6);
    EXPECT_TRUE(output.valid);
}

TEST_F(ExampleServiceTest, ZeroInput) {
    model::InputData input{.value = 0.0, .timestamp = 0.0};
    auto output = service_->process(input);
    EXPECT_NEAR(output.result, 0.0, 1e-6);
    EXPECT_FALSE(output.valid);
}

TEST_F(ExampleServiceTest, NegativeInputThrows) {
    model::InputData input{.value = -1.0, .timestamp = 0.0};
    EXPECT_THROW(service_->process(input), model::InputValidationError);
}

TEST_F(ExampleServiceTest, ConfigUpdate) {
    model::Config new_config;
    new_config.threshold = 2.0;
    service_->update_config(new_config);

    model::InputData input{.value = 3.0, .timestamp = 0.0};
    auto output = service_->process(input);
    EXPECT_NEAR(output.result, 6.0, 1e-6);
}

File: templates/new_node_py/test/test_service.py
# ═══ 纯逻辑测试：不需要 ROS2，直接测试 Service 层 ═══
"""ExampleService 的 pytest 测试。

不启动任何 ROS2 节点，只测 service 层纯逻辑。
"""
import pytest

from __PACKAGE_NAME__.model import Config, InputData, InputValidationError
from __PACKAGE_NAME__.service import ExampleService


@pytest.fixture
def svc() -> ExampleService:
    return ExampleService(Config(threshold=0.5, update_rate=10.0))


def test_normal_input(svc: ExampleService) -> None:
    out = svc.process(InputData(value=2.0, timestamp=0.0))
    assert out.result == pytest.approx(1.0)
    assert out.valid is True


def test_zero_input(svc: ExampleService) -> None:
    out = svc.process(InputData(value=0.0, timestamp=0.0))
    assert out.result == pytest.approx(0.0)
    assert out.valid is False


def test_negative_input_raises(svc: ExampleService) -> None:
    with pytest.raises(InputValidationError):
        svc.process(InputData(value=-1.0, timestamp=0.0))


def test_config_update(svc: ExampleService) -> None:
    svc.update_config(Config(threshold=2.0, update_rate=10.0))
    out = svc.process(InputData(value=3.0, timestamp=0.0))
    assert out.result == pytest.approx(6.0)
